File: src/main/java/io/trydent/olimpo/sink/Command.java
package io.trydent.olimpo.sink;

import io.vertx.core.json.JsonObject;

import java.util.function.Consumer;

@FunctionalInterface
public interface Command extends Consumer<JsonObject> {
  @Override
  default void accept(JsonObject params) {
    this.execute(params);
  }

  void execute(JsonObject params);
}

File: src/main/java/io/trydent/olimpo/dispatch/Command.kt
package io.trydent.olimpo.dispatch

import io.trydent.olimpo.apollo.Id
import io.trydent.olimpo.http.media.Json
import io.vertx.core.eventbus.EventBus
import java.util.UUID.randomUUID

class CommandId : Id {
  override fun invoke() = "${randomUUID()}"
}

interface Command : (String, Json) -> Id
interface Event : (String, Json) -> Unit

class BusCommand(private val bus: EventBus) : Command {
  override fun invoke(name: String, params: Json): Id {
    bus.publish("$name-command", params)
    return CommandId()
  }
}
